pkg/kubelet/kubelet.go

```diff
@@ -1266,6 +1266,67 @@ func (kl *Kubelet) setupDataDirs() error {
 	return nil
 }
 
+// chownDirForRemappedIDs change dir and path ownerships if UserNamespace is enabled at container runtime
+// noop if UserNamespace is disabled
+func (kl *Kubelet) chownDirForRemappedIDs(path string) error {
+	isUseNamespaceSupportedAndEnabled, err := kl.isUserNamespaceRemappingEnabledAtRuntime()
+	if err != nil {
+		klog.Warningf("error while determining usernamespace configuration at runtime: %v", err)
+		return nil
+	}
+	if !isUseNamespaceSupportedAndEnabled {
+		// No-Op if UserNamespace remapping is not enabled at runtime
+		return nil
+	}
+	if path == "" {
+		return fmt.Errorf("path to be setup is empty")
+	}
+	if err := kl.chownAllFilesAt(path); err != nil {
+		return fmt.Errorf("error setting ownership to remapped IDs on %s: %v", path, err)
+	}
+	return nil
+}
+
+// chownAllFilesAt traverses the directory tree at the give path and chowns the paths to adjust for the remapped usernamespaces
+func (kl *Kubelet) chownAllFilesAt(dir string) error {
+	var files []string
+
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		files = append(files, path)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	klog.V(5).Infof("Chowned paths %v", files)
+	for _, file := range files {
+		_, fileGID, err := kl.getOwnerIDsFor(file)
+		if err != nil {
+			return fmt.Errorf("error in getting volume path owner UID/GID: %v", err)
+		}
+		if fileGID != 0 {
+			klog.V(5).Infof("GID, %v, for path %v is not equal to 0. Skipping chowing assuming it to be FsGroup GID ", fileGID, file)
+			continue
+		}
+		containerUID := 0
+		containerGID := 0
+		uid, err := kl.getHostUID(containerUID)
+		if err != nil {
+			return fmt.Errorf("Failed to get remapped host UID corresponding to UID 0 in container namespace: %v", err)
+		}
+		gid, err := kl.getHostGID(containerGID)
+		if err != nil {
+			return fmt.Errorf("Failed to get remapped host GID corresponding to GID 0 in container namespace: %v", err)
+		}
```

Comment:
I don't understand this logic. Why to calculate it each time if container is always 0 and the result should be the same?

Reply:
`containerUID` is zero (root in the userns) but `uid` is the mapping on the host (e.g. 100000). So we correctly do a `os.Chown(file, 100000, 100000)` below.

Reply:
What I meant is that the calculation is inside the loop but IIUC it's the same for all the files, so we could do it once.